ansible_base/models/rbac.py

```diff
@@ -0,0 +1,389 @@
+
+        if actor._meta.model_name == 'user':
+            if giving:
+                object_role.users.add(actor)
+            else:
+                object_role.users.remove(actor)
+        elif isinstance(actor, permission_registry.team_model):
+            if giving:
+                object_role.teams.add(actor)
+            else:
+                object_role.teams.remove(actor)
+        else:
+            raise RuntimeError(f'Cannot give permission to {actor}, must be a user or team')
+
+        if (not giving) and (not (object_role.users.exists() or object_role.teams.exists())):
+            if object_role in to_update:
+                to_update.remove(object_role)
+            object_role.delete()
+
+        update_after_assignment(update_teams, to_update)
+
+        if not sync_action and self.name in permission_registry._trackers:
+            tracker = permission_registry._trackers[self.name]
+            with tracker.sync_active():
+                tracker.sync_relationship(actor, content_object, giving=giving)
+
+        return object_role
+
+
+class ObjectRole(models.Model):
+    """
+    This is the successor to the Role model in the old AWX RBAC system
+    It is renamed to ObjectRole to distinguish from the abstract or generic
+    RoleDefinition which does not apply to a particular object.
+
+    This matches the RoleDefinition to a content_object.
+    After this is created, users and teams can be added to gives those
+    permissions to that user or team, for that content_object
+    """
+
+    class Meta:
+        app_label = 'ansible_base'
+        verbose_name_plural = _('object_roles')
+        indexes = [models.Index(fields=["content_type", "object_id"])]
+        ordering = ("content_type", "object_id")
+        constraints = [models.UniqueConstraint(name='one_object_role_per_object_and_role', fields=['object_id', 'content_type', 'role_definition'])]
+
+    role_definition = models.ForeignKey(
+        RoleDefinition,
+        on_delete=models.CASCADE,
+        help_text=_("The role definition which defines what permissions this object role grants"),
+    )
+
+    users = models.ManyToManyField(
+        settings.AUTH_USER_MODEL, related_name='has_roles', help_text=_("Users who have access to the permissions defined by this object role")
+    )
+    teams = models.ManyToManyField(
+        settings.ROLE_TEAM_MODEL, related_name='has_roles', help_text=_("Teams or groups who have access to the permissions defined by this object role")
+    )
+    # COMPUTED DATA
+    provides_teams = models.ManyToManyField(
+        settings.ROLE_TEAM_MODEL,
+        related_name='member_roles',
+        help_text=_("Users who have this role obtain member access to these teams, and inherit all their permissions"),
+    )
+
+    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE)
+    object_id = models.PositiveIntegerField()
```

Comment:
TIL

https://docs.djangoproject.com/en/5.0/ref/contrib/contenttypes/

> The “object_id” field doesn’t have to be the same type as the primary key fields on the related models, but their primary key values must be coercible to the same type as the “object_id” field by its [get_db_prep_value()](https://docs.djangoproject.com/en/5.0/ref/models/fields/#django.db.models.Field.get_db_prep_value) method.

As these docs continue elaborating, they seem to point in the direction of CharField for our needs. As you say.

Reply:
The main concern for me is the UUID pks in Hub and EDA

Reply:
This has been updated. I will work on adding a docs section for it too. Since the Resource API merged its methods are accepted as canonical to some extent. The _specification_ models now use TextField in the same way as the Resource API. This is important so that coherent system-wide lists can be presented to the user. The _evaluation_ model now maintains different tables for pk and uuid resources.